File: python/question1_tests.py
"""
Test infrastructure for Problem 1 -- Cow Sighting Frequency.
Contains reference solution, input generators, and all test cases.
"""

import random
from collections import Counter
from test_harness import TestHarness, Case


def build_harness(solve_fn):
    """Build a fully configured test harness for Problem 1."""
    stress_rng = random.Random(123)
    large_input = _gen(200_000, 1_000_000)

    return TestHarness("Cow Sighting Frequency") \
        .add_test("Basic test",
            lambda: solve_fn([1, 2, 1, 3, 1]),
            (1, 3)) \
        .add_test("Single element",
            lambda: solve_fn([5]),
            (5, 1)) \
        .add_test("All same",
            lambda: solve_fn([42, 42, 42]),
            (42, 3)) \
        .add_test("Tie -> smallest ID wins",
            lambda: solve_fn([3, 3, 2, 2]),
            (2, 2)) \
        .add_test("Large IDs",
            lambda: solve_fn([999_999, 1_000_000, 999_999]),
            (999_999, 2)) \
        .add_test("Edge: two elements, different, ascending",
            lambda: solve_fn([1, 2]),
            (1, 1)) \
        .add_test("Edge: two elements, different, descending",
            lambda: solve_fn([2, 1]),
            (1, 1)) \
        .add_test("Edge: max ID only",
            lambda: solve_fn([1_000_000]),
            (1_000_000, 1)) \
        .add_scaling("N (array size, maxId = 1,000,000)",
            [100, 1_000, 10_000, 50_000, 100_000, 200_000],
            lambda n: _gen(n, 1_000_000),
            lambda inp: solve_fn(inp),
            lambda inp: _brute_force(inp)) \
        .add_random_tests("Stress test (N=1..500, maxId=1..100)", 1_000,
            lambda: (
                lambda n=stress_rng.randint(1, 500), mx=stress_rng.randint(1, 100):
                    (lambda inp=_gen(n, mx, stress_rng.randint(0, 10**6)):
                        Case(f"N={n} maxId={mx}", lambda i=inp: solve_fn(i[:]), _brute_force(inp))
                    )()
            )()) \
        .add_benchmark("Large random (N = 200,000)",
            lambda: solve_fn(large_input[:]))


# ── Reference solution ───────────────────────────────────

def _brute_force(sightings):
    counts = Counter(sightings)
    best_id, best_freq = -1, 0
    for id_, freq in counts.items():
        if freq > best_freq or (freq == best_freq and id_ < best_id):
            best_id, best_freq = id_, freq
    return (best_id, best_freq)


# ── Input generators ─────────────────────────────────────

def _gen(n, max_id, seed=42):
    rng = random.Random(seed)
    return [rng.randint(1, max_id) for _ in range(n)]

File: java/Question1_Tests.java
import java.util.*;
import java.util.function.*;

/**
 * Test infrastructure for Problem 1 — Cow Sighting Frequency.
 * Contains reference solution, input generators, and all test cases.
 */
class Question1_Tests {

    static TestHarness<int[]> buildHarness(Function<int[], int[]> solveFn) {
        Random rng = new Random(123);
        int[] largeInput = generateRandomInput(200_000, 1_000_000);

        return TestHarness.<int[]>forProblem("Cow Sighting Frequency")

            // ── Correctness tests ────────────────────────────
            .addTest("Basic test",
                () -> solveFn.apply(new int[]{1, 2, 1, 3, 1}),
                new int[]{1, 3})

            .addTest("Single element",
                () -> solveFn.apply(new int[]{5}),
                new int[]{5, 1})

            .addTest("All same",
                () -> solveFn.apply(new int[]{42, 42, 42}),
                new int[]{42, 3})

            .addTest("Tie \u2192 smallest ID wins",
                () -> solveFn.apply(new int[]{3, 3, 2, 2}),
                new int[]{2, 2})

            .addTest("Large IDs",
                () -> solveFn.apply(new int[]{999_999, 1_000_000, 999_999}),
                new int[]{999_999, 2})

            .addTest("Edge: two elements, different, ascending",
                () -> solveFn.apply(new int[]{1, 2}),
                new int[]{1, 1})

            .addTest("Edge: two elements, different, descending",
                () -> solveFn.apply(new int[]{2, 1}),
                new int[]{1, 1})

            .addTest("Edge: max ID only",
                () -> solveFn.apply(new int[]{1_000_000}),
                new int[]{1_000_000, 1})

            // ── Scaling analysis ─────────────────────────────
            .addScaling("N (array size, maxId = 1,000,000)",
                new int[]{100, 1_000, 10_000, 50_000, 100_000, 200_000},
                n -> generateRandomInput(n, 1_000_000),
                input -> solveFn.apply(input),
                input -> bruteForce(input))

            // ── Random stress tests ──────────────────────────
            .addRandomTests("Stress test (N=1..500, maxId=1..100)", 1_000, () -> {
                int n = rng.nextInt(500) + 1;
                int maxId = rng.nextInt(100) + 1;
                int[] input = generateRandomInput(n, maxId, rng.nextInt());
                int[] expected = bruteForce(input);
                return new TestHarness.Case<>(
                    "N=" + n + " maxId=" + maxId,
                    () -> solveFn.apply(input.clone()),
                    expected);
            })

            // ── Benchmark ────────────────────────────────────
            .addBenchmark("Large random (N = 200,000)",
                () -> solveFn.apply(largeInput.clone()));
    }

    // ── Reference solution ───────────────────────────────────

    private static int[] bruteForce(int[] sightings) {
        Map<Integer, Integer> counts = new HashMap<>();
        for (int id : sightings) {
            counts.merge(id, 1, Integer::sum);
        }
        int bestId = -1, bestFreq = 0;
        for (Map.Entry<Integer, Integer> entry : counts.entrySet()) {
            int id = entry.getKey(), freq = entry.getValue();
            if (freq > bestFreq || (freq == bestFreq && id < bestId)) {
                bestId = id;
                bestFreq = freq;
            }
        }
        return new int[]{bestId, bestFreq};
    }

    // ── Input generators ─────────────────────────────────────

    private static int[] generateRandomInput(int n, int maxId) {
        return generateRandomInput(n, maxId, 42);
    }

    private static int[] generateRandomInput(int n, int maxId, int seed) {
        Random rng = new Random(seed);
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = rng.nextInt(maxId) + 1;
        }
        return arr;
    }
}
